Add visualization for RtDose files.

Kikooku opened the pull request:
#### What contibution are you adding? How do we use it?

A python script that you can run with spyder3 to visualize RtDose files.

 - [X] I have added a single script or subfolder to a topic folder
 - [X] I have added some form of documentation and my contact with the contribution
 - [X] I have added a link to the contribution in the topic folder's `README.md`

Kikooku committed:
Fix problem on three viewer

I've had problem (compat.in_py2 not existing) running the program in windows. Here is an update that should fix it.

Kikooku committed:
Create rtdose_visual.py

I've created a python script that allow us to visualize Rt Dose DICOM files in three 2D plots.

Kikooku committed:
Update README.md

darcymason commented on plotting-visualization/dicomtree.py:
```diff
@@ -15,9 +15,12 @@
+        import Tix as tkinter_tix
+    else:
+        import tkinter.tix as tkinter_tix
+except:
```
Better to use `except ImportError` to avoid catching other errors that might arise

ferdymercury replied:
It should be instead `except AttributeError:`

to avoid this:
`AttributeError: module 'pydicom.compat' has no attribute 'in_py2'`

darcymason replied:
Oh, that is because pydicom no longer supports Python 2, so the `in_py2` flag has been removed.  The `contrib` scripts should by now remove all references to `in_py2` and assume pydicom >= 2.0.

darcymason commented on plotting-visualization/rtdose_visual.py:
```diff
@@ -0,0 +1,199 @@
+   
+def multi_slice_viewer(volume):
+    """
+    Based on : https://www.datacamp.com/community/tutorials/matplotlib-3d-volumetric-data
```
What is the licensing on that code?  If not MIT-compatible, then should not be using any code directly (not sure what "based on" implies here).

ferdymercury replied:
It's not a source code directly, it's just some snippets within a Python visualization tutorial. So not sure if a license applies at all. They do not say anything explicitly in that post.

With "Based" we mean that some of these functions are essentially the same, with some minor modifications to allow changing the axis of view, not just the slice number.

darcymason replied:
Can you check a little further?  I think a function that is essentially the same would require the license to be compatible with pydicom.  The repo at https://github.com/datacamp/datacamp-community-tutorials, with a few tutorials, is MIT-license, but I don't know if that implies that all are.  I would expect datacamp should have a statement somewhere about tutorial code license if not otherwise specified, but perhaps the author can be contacted if not.

ferdymercury replied:
Thanks for the directions. I will contact the author @jni.

ferdymercury replied:
He answered, he says the original code is available here https://github.com/jni/mpl-volume-viewer/blob/master/LICENSE under a BSD-3, so he said we can reuse it without problems.

jni replied:
👋 Hello! I responded to @ferdymercury privately but since I got the link here:

The code is fully available here:

https://github.com/jni/mpl-volume-viewer

and is BSD 3-clause. Enjoy! ;)

darcymason replied:
@kikooku or @ferdymercury, I suggest then that the "based on" link in this PR's code be updated to point to the BSD location.  With that and removal of `in_py2`, we can get this merged in.

ferdymercury replied:
We'll do that, thanks. We are cleaning up a couple of things in the code also.

Kikooku replied:
Done

darcymason reviewed (changes requested):
Thanks for this, just one adjustment and a question

Kikooku committed:
Optimization of the program

A version that reduces the use of global variables and reestructurates some of the functions. It is still incomplete when reading the file name instead of folder name

Kikooku committed:
Update rtdose_visual.py

A quick fix that allows the user to chose between giving an input of the name of the file or the name of the folder the file is in.

Kikooku committed:
Update rtdose_visual.py

Kikooku committed:
Update dicomtree.py

Eliminate the try method that gave was implemented previosly

darcymason commented on plotting-visualization/dicomtree.py:
```diff
@@ -14,14 +14,7 @@
 from __future__ import print_function
```
If Python 3 only, then don't need the `__future__` import.

Kikooku replied:
Done

darcymason commented on plotting-visualization/rtdose_visual.py:
```diff
@@ -0,0 +1,250 @@
+    ax.axis = (ax.axis + 1) % 3
+    ax.index = volume.shape[ax.axis] // 2
+
+# Define the path folder where the file is , or directly the RTdose full file name
```
I suggest that the top-level code be indented under a `if __name__ == "__main__":`, so functions could be imported and used from other code.

Kikooku replied:
Done

darcymason commented on plotting-visualization/rtdose_visual.py:
```diff
@@ -0,0 +1,250 @@
+    dirs = os.listdir( path )
+    for file in dirs:
+       ds = pydicom.dcmread(path+"/"+file)
+       if ds.file_meta.MediaStorageSOPClassUID == '1.2.840.10008.5.1.4.1.1.481.2':
```
Here `file` could be itself be a directory, leading to an error.  Could trap the exception or embed in an `if not os.path.isdir(file):`...

darcymason replied:
Similarly, without the `force=True` passed to `dcmread`, this will bail with an error on any file that it not actually DICOM, or is old-style DICOM without proper preamble, etc. (like pydicom's rtdose.dcm test file).

Kikooku replied:
Done

darcymason reviewed (commented):
Apologies for adding more content when I implied before that we were essentially done, but this time I checked out the PR and tried running it, which exposed some more problems.

Just as a style request, it would be good to run the code through a formatter like `black` to fix non pep-8 style like missing spaces around commas and operators.

Also, I also was unable to get the visualization in both Windows and linux running from command line using a couple of different test files (including pydicom's `rtdose.dcm` test file).  Min/max/median etc. were shown, however.  Perhaps it is something to do with my setup, but unfortunately no error is thrown, just the message "If the program has not plotted the scan, check if there is an RTDose file in the path folder".  Any ideas?  I just installed `matplotlib`, nothing else.

ferdymercury commented:
>  "If the program has not plotted the scan, check if there is an RTDose file in the path folder". Any ideas? I just installed `matplotlib`, nothing else.

Did you run it with Spyder3? Otherwise the plot does not show up.

If run through terminal, try to add:
`plt.show()` so that it appears.

We'll take care of the other stuff.

darcymason commented:
> If run through terminal, try to add:
> `plt.show()` so that it appears.

Yes, that worked, thanks.  But the next/prev etc. keys do not work   However, it's not really necessary to get this command line method working if you add some comments at the beginning that it is only for the Spyder environment.

Kikooku committed:
Remove python2 command

Kikooku committed:
Formatted by black, if__name__ implemented and problems with reading files

Kikooku committed:
Fixed missing path + "/"

Kikooku committed:
Added a comment about the spyder environment

Kikooku committed:
Erase import compat and simplify the if

Kikooku committed:
Merge branch 'master' into master

darcymason reviewed (approved):
Looks good - just the one typo

ferdymercury commented:
> Looks good - just the one typo

thanks! which typo do you mean?

darcymason commented:
> thanks! which typo do you mean?

Oops - that got lost somehow.  Very minor, just in a comment line, to replace "gona" with "going to".

ferdymercury commented on plotting-visualization/rtdose_visual.py:
```diff
@@ -0,0 +1,265 @@
+#!/usr/bin/python
+
+
+# Here we define the functions that we are gona use in our program
```
@Kikooku as Darcy suggested, please change "gona" with "going to". Thanks!!

Kikooku replied:
Done!

Kikooku committed:
Fixed typo

Kikooku committed:
Merge branch 'master' of https://github.com/Kikooku/contrib-pydicom
